testing/trino-product-tests-launcher/src/main/java/io/trino/tests/product/launcher/cli/TestRun.java

```diff
@@ -219,19 +244,57 @@ public Integer call()
 
         private Integer tryExecuteTests()
         {
-            try (Environment environment = startEnvironment()) {
-                return toIntExact(environment.awaitTestsCompletion());
+            Environment environment = getEnvironment();
+            if (!hasImpactedFeatures(environment)) {
+                log.warn("Skipping test due to impacted features not overlapping with any features configured in environment");
+                return toIntExact(ENVIRONMENT_SKIPPED_EXIT_CODE);
+            }
+            try (Environment runningEnvironment = startEnvironment(environment)) {
+                return toIntExact(runningEnvironment.awaitTestsCompletion());
             }
             catch (RuntimeException e) {
                 log.warn(e, "Failed to execute tests");
                 return ExitCode.SOFTWARE;
             }
         }
 
-        private Environment startEnvironment()
+        private boolean hasImpactedFeatures(Environment environment)
```

Comment:
What's a "feature" in this context? How do we track them, keep them up to date, etc?

Reply:
Please read the PR description that explains this and also previous comments, like https://github.com/trinodb/trino/pull/10895#issuecomment-1059520957

Reply:
I saw:

> map modules to plugins and extract features they provide (like connector names, password authorization etc)

But I still don't understand how we maintain that mapping.

Reply:
The plugin interface has a method for each feature. I wrote a plugin reader tool that lists every feature provided by every plugin. It's similar to what the plugin manager does. If there would be a new kind of feature added to this interface, it should be handled in both the plugin manager and plugin reader.

Reply:
@Ashhar suggested adding a test that would use reflection to make sure all plugin features are covered in the plugin reader. Do you think adding such a test is required to merge this?